src/numeric/decision_rules.jl

```diff
@@ -17,190 +51,169 @@ function filter_mcoeffs(a::Array{Float64,1},b::Array{Float64,1},n::Array{Int64,1
-    n_x::Int # number of values
-    coefficients::Array{Array{Float64}}
+function DecisionRule(grid_exo::EmptyGrid, grid_endo::CartesianGrid, values::Array{Array{Float64,2}})
+    n_x = size(values[1], 2)
+    dr = DecisionRule(grid_exo, grid_endo, n_x)
+    set_values!(dr, values)
+    return dr
 end
 
-function DecisionRule(mc::DiscreteMarkovProcess, grid::CartesianGrid, n_x::Int)
-    a = grid.min
-    b = grid.max
-    orders = grid.n
-    n_ms = size(mc.values, 1)
-    coeffs = [zeros(n_x, (orders+2)...) for v in 1:n_ms]
-    return MCDecisionRule(mc, grid, n_x,  coeffs)
+function set_values!(dr::AbstractDecisionRule{EmptyGrid, CartesianGrid}, values::Array{Array{Float64,2},1})
+    a = dr.grid_endo.min
+    b = dr.grid_endo.max
+    orders = dr.grid_endo.n
+    dr.coefficients = [filter_mcoeffs(a, b, orders, v) for v in values]
 end
 
-function DecisionRule(mc::DiscreteMarkovProcess, grid::CartesianGrid, values::Array{Array{Float64,2}})
-    n_x = size(values[1], 2)
-    dr = DecisionRule(mc, grid, n_x)
-    set_values!(dr, values)
-    return dr
+function set_values!(dr::AbstractDecisionRule{EmptyGrid, CartesianGrid}, values::Array{Float64,2})
+    set_values!(dr, [values])
 end
 
-function (dr::MCDecisionRule)(i::Int, x::Array{Float64,2})
-    a = dr.grid.min
-    b = dr.grid.max
-    n = dr.grid.n
-    cc = dr.coefficients[i]
-    res = splines.eval_UC_multi_spline(a, b, n, cc, x)'
+function evaluate(dr::AbstractDecisionRule{EmptyGrid, CartesianGrid},z::Matrix{Float64})
+    a = dr.grid_endo.min
+    b = dr.grid_endo.max
+    n = dr.grid_endo.n
+    cc = dr.coefficients[1]
+    res = splines.eval_UC_multi_spline(a, b, n, cc, z)'
     return res
 end
 
-(dr::MCDecisionRule)(i::Int, x::Array{Float64,1}) =  dr(i, x')[:]
-(dr::MCDecisionRule)(i::Int, j::Int, x::Union{Vector, Matrix}) = dr(j, x)
+(dr::DecisionRule{EmptyGrid, CartesianGrid})(z::Matrix{Float64}) = evaluate(dr,z)
+(dr::DecisionRule{EmptyGrid, CartesianGrid})(z::Vector{Float64}) = dr(z')[:]
+(dr::DecisionRule{EmptyGrid, CartesianGrid})(i::Int64, x::Union{Vector{Float64},Matrix{Float64}}) = dr(x)
+(dr::DecisionRule{EmptyGrid, CartesianGrid})(x::Union{Vector{Float64},Matrix{Float64}}, y::Union{Vector{Float64},Matrix{Float64}}) = dr(y)
 
 
-# Decision Rule on IID Process
+####
+#### 2 continous arguments d.r.
+####
 
-type IDecisionRule <: AbtsractCubicSplineDR
-    process::IIDExogenous
-    grid::CartesianGrid
-    n_x::Int # number of values
-    coefficients::Vector{Array{Float64}}
-end
-
-function DecisionRule(proc::MvNormal, grid::CartesianGrid, n_x::Int)
-    a = grid.min
-    b = grid.max
-    orders = grid.n
+function DecisionRule(grid_exo::CartesianGrid, grid_endo::CartesianGrid, n_x::Int)
+    # hmm kind of silently assuming we have cartesian grid
+    orders = grid_exo.n + grid_endo.n
     coeffs = [zeros(n_x, (orders+2)...)]
-    return IDecisionRule(proc, grid, n_x,  coeffs)
+    return (DecisionRule{CartesianGrid, CartesianGrid})(grid_exo, grid_endo, n_x, coeffs)
```

Comment:
Specifying the type parameters here is also not needed.

This should be written

```julia
DecisionRule(grid_exo, grid_endo, n_x, coefs)
```